File: benchmark/lua/power.lua
local BASE = 1000000000

local function add_big(a, b)
    local n = #a
    if #b > n then
        n = #b
    end
    local result = {}
    local carry = 0
    for i = 1, n do
        local sum = (a[i] or 0) + (b[i] or 0) + carry
        if sum >= BASE then
            sum = sum - BASE
            carry = 1
        else
            carry = 0
        end
        result[i] = sum
    end
    if carry > 0 then
        result[#result + 1] = carry
    end
    return result
end

local function mul_small(a, n)
    local result = {}
    local carry = 0
    for i = 1, #a do
        local prod = a[i] * n + carry
        result[i] = prod % BASE
        carry = prod // BASE
    end
    while carry > 0 do
        result[#result + 1] = carry % BASE
        carry = carry // BASE
    end
    return result
end

local function big_to_string(a)
    local parts = {tostring(a[#a])}
    for i = #a - 1, 1, -1 do
        parts[#parts + 1] = string.format("%09d", a[i])
    end
    return table.concat(parts)
end

local function power(base, exp)
    local result = {1}
    for i = 1, exp do
        result = mul_small(result, base)
    end
    return result
end

local start = os.clock()
local sum = {0}
for i = 1, 10000 do
    sum = add_big(sum, power(2, 1000))
end
local finish = os.clock()
print("Sum: " .. big_to_string(sum))
print(string.format("Time: %.0fms", (finish - start) * 1000))

File: benchmark/perl/power.pl
use Time::HiRes qw(time);
use constant BASE => 1000000000;

sub add_big {
    my ($a, $b) = @_;
    my $n = @$a > @$b ? @$a : @$b;
    my @result;
    my $carry = 0;
    for (my $i = 0; $i < $n; $i++) {
        my $sum = ($a->[$i] // 0) + ($b->[$i] // 0) + $carry;
        if ($sum >= BASE) {
            $sum -= BASE;
            $carry = 1;
        } else {
            $carry = 0;
        }
        $result[$i] = $sum;
    }
    push @result, $carry if $carry > 0;
    return \@result;
}

sub mul_small {
    my ($a, $n) = @_;
    my @result;
    my $carry = 0;
    for (my $i = 0; $i < @$a; $i++) {
        my $prod = $a->[$i] * $n + $carry;
        $result[$i] = $prod % BASE;
        $carry = int($prod / BASE);
    }
    while ($carry > 0) {
        push @result, $carry % BASE;
        $carry = int($carry / BASE);
    }
    return \@result;
}

sub big_to_string {
    my ($a) = @_;
    my $result = "$a->[-1]";
    for (my $i = @$a - 2; $i >= 0; $i--) {
        $result .= sprintf("%09d", $a->[$i]);
    }
    return $result;
}

sub power {
    my ($base, $exp) = @_;
    my $result = [1];
    for (my $i = 0; $i < $exp; $i++) {
        $result = mul_small($result, $base);
    }
    return $result;
}

my $start = time * 1000;
my $sum = [0];
for (1..10000) {
    $sum = add_big($sum, power(2, 1000));
}
my $end = time * 1000;
print "Sum: " . big_to_string($sum) . "\n";
printf "Time: %.0fms\n", $end - $start;
